core/shared/src/main/scala/spire/math/interval/Overlap.scala

```diff
@@ -0,0 +1,89 @@
+package spire.math.interval
+
+import spire.algebra.{Eq, Order}
+import spire.math.{Bounded, Empty, Interval, Point}
+import spire.math.interval.Overlap.{Disjoint, Equal, Subset}
+import spire.syntax.eq._
+/**
+  * An ADT that represents overlapping result for any two intervals.
+  */
+sealed abstract class Overlap[A] extends Product with Serializable {
+  def isDisjoint: Boolean = this.isInstanceOf[Disjoint[_]]
+  def isSubset: Boolean = this match {
+    case Subset(_, _) | Equal() => true
+    case _ => false
+  }
+  def isEqual: Boolean = this.isInstanceOf[Equal[_]]
+}
+
+object Overlap {
+  /**
+    * Intervals are nonEmpty and don't intersect
+    * [[lower.upperBound]] is strictly less than [[upper.lowerBound]].
+    */
+  case class Disjoint[A] private[spire](lower: Interval[A], upper: Interval[A]) extends Overlap[A] {
+
+    /**
+      * An interval that joins [[lower]] and [[upper]] in a continuous interval without intersecting any of them.
+      * For example for (-5, 1] and (4, 6), a join is (1,4]
+      */
+    def join: Interval[A] = (~lower).last.intersect((~upper).head)
+  }
+
+  /**
+    * Non empty intervals, for which holds:
+    * [[upper]] ∋ [[lower.upperBound]] && [[upper]] ∌ [[lower.lowerBound]]
+    * For example: (-2, 10] and [5, 13)
+    */
+  case class PartialOverlap[A] private[spire](lower: Interval[A], upper: Interval[A]) extends Overlap[A]
+
+  /**
+    * [[inner]] is a subset of [[outer]].
+    * Empty interval is always a subset of any other, so all overlaps on empty intervals go here,
+    * except `(Ø).overlap(Ø)`, that results in equality.
+    *
+    * For example [1,4) and [1, 5]
+    */
+  case class Subset[A] private[spire](inner: Interval[A], outer: Interval[A]) extends Overlap[A]
+
+  /**
+    * Intervals are equal
+    */
+  case class Equal[A] private[spire]() extends Overlap[A]
+
+
+  def apply[A: Order](lhs: Interval[A], rhs: Interval[A]): Overlap[A] = {
+
+    def lessAndOverlaps(intersectionLowerBound: Bound[A]): Overlap[A] =
+      if (lhs.lowerBound === intersectionLowerBound) PartialOverlap(lhs, rhs) else PartialOverlap(rhs, lhs)
+
+    if (lhs === rhs) {
+      Equal()
+    } else {
+      (lhs, rhs) match {
+        case (sub, sup) if sup.isSupersetOf(sub) => Subset(sub, sup)
+        case (sup, sub) if sup.isSupersetOf(sub) => Subset(sub, sup)
```

Comment:
I think this outer pattern match could be better-written as:

```scala
if (rhs.isSupersetOf(lhs)) Subset(lhs, rhs)
else if (lhs.isSupersetOf(rhs)) Subset(rhs, lhs)
else {
  ...
}
```

It seems a bit nit-picky, but when I was reviewing this code I missed the fact that `sub` and `sup` were switched in the two cases, and thought the code was the same. (Those names are also very similar so I think it's easy to get confused.) It doesn't seem like the pattern-match is saving us any code, so maybe we can just use `lhs` and `rhs` with if/else branches?